sdks/v4-sdk/src/entities/position.ts

```diff
@@ -0,0 +1,194 @@
+// @ts-nocheck
+import { BigintIsh, MaxUint256, Percent, Price, CurrencyAmount, Currency } from '@uniswap/sdk-core'
+import JSBI from 'jsbi'
+import invariant from 'tiny-invariant'
+import { ZERO } from '../internalConstants'
+import { maxLiquidityForAmounts } from '../utils/maxLiquidityForAmounts'
+import { tickToPrice } from '../utils/priceTickConversions'
+import { SqrtPriceMath } from '../utils/sqrtPriceMath'
+import { TickMath } from '../utils/tickMath'
+import { encodeSqrtRatioX96 } from '../utils/encodeSqrtRatioX96'
+import { Pool } from './pool'
+
+interface PositionConstructorArgs {
+  pool: Pool
+  liquidity: BigintIsh
+  tickLower: number
+  tickUpper: number
+}
+
+/**
+ * Represents a position on a Uniswap V4 Pool
+ * @dev Similar to the V3 implementation
+ * - using Currency instead of Token
+ * - keep in mind that Pool and liquidity must be fetched from the pool manager
+ */
+export class Position {
+  public readonly pool: Pool
+  public readonly tickLower: number
+  public readonly tickUpper: number
+  public readonly liquidity: JSBI // TODO: needs to be fetched from pool manager
+
+  // cached resuts for the getters
+  private _token0Amount: CurrencyAmount<Currency> | null = null
+  private _token1Amount: CurrencyAmount<Currency> | null = null
+  private _mintAmounts: Readonly<{ amount0: JSBI; amount1: JSBI }> | null = null
+
+  /**
+   * Constructs a position for a given pool with the given liquidity
+   * @param pool For which pool the liquidity is assigned
+   * @param liquidity The amount of liquidity that is in the position
+   * @param tickLower The lower tick of the position
+   * @param tickUpper The upper tick of the position
+   */
+  public constructor({ pool, liquidity, tickLower, tickUpper }: PositionConstructorArgs) {
+    invariant(tickLower < tickUpper, 'TICK_ORDER')
+    invariant(tickLower >= TickMath.MIN_TICK && tickLower % pool.tickSpacing === 0, 'TICK_LOWER')
+    invariant(tickUpper <= TickMath.MAX_TICK && tickUpper % pool.tickSpacing === 0, 'TICK_UPPER')
+
+    this.pool = pool
+    this.tickLower = tickLower
+    this.tickUpper = tickUpper
+    this.liquidity = JSBI.BigInt(liquidity)
+  }
+
+  /**
+   * Returns the price of token0 at the lower tick
+   */
+  public get token0PriceLower(): Price<Currency, Currency> {
+    // TODO: Currency or wrapped token here?
+    throw new Error('Not implemented')
+  }
+
+  /**
+   * Returns the price of token0 at the upper tick
+   */
+  public get token0PriceUpper(): Price<Currency, Currency> {
+    throw new Error('Not implemented')
+  }
+
+  /**
+   * Returns the amount of token0 that this position's liquidity could be burned for at the current pool price
+   */
+  public get amount0(): CurrencyAmount<Currency> {
```

Comment:
doesn't this calculation need the current price and price at lower and price at upper?

Reply:
discussed offline - will do similar to v3 impl.